pkg/storage/redis/redis.go

```diff
@@ -0,0 +1,176 @@
+package redis
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/gomodule/redigo/redis"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	internal_databroker "github.com/pomerium/pomerium/internal/databroker"
+	"github.com/pomerium/pomerium/pkg/grpc/databroker"
+	"github.com/pomerium/pomerium/pkg/storage"
+)
+
+var _ storage.Backend = (*DB)(nil)
+
+// DB wraps redis conn to interact with redis server.
+type DB struct {
+	pool                   *redis.Pool
+	deletePermanentlyAfter int64
+	recordType             string
+	lastVersion            uint64
+}
+
+// New returns new DB instance.
+func New(address string) (*DB, error) {
+	db := &DB{
+		pool: &redis.Pool{
+			Wait: true,
+			DialContext: func(ctx context.Context) (redis.Conn, error) {
+				ctx, cancelFn := context.WithTimeout(ctx, 5*time.Second)
+				defer cancelFn()
+				c, err := redis.DialContext(ctx, "tcp", address)
+				if err != nil {
+					return nil, fmt.Errorf(`redis.DialURL(): %w`, err)
+				}
+				return c, nil
+			},
+			TestOnBorrow: func(c redis.Conn, t time.Time) error {
+				if time.Since(t) < time.Minute {
+					return nil
+				}
+				_, err := c.Do("PING")
+				if err != nil {
+					return fmt.Errorf(`c.Do("PING"): %w`, err)
+				}
+				return nil
+			},
+		},
+		deletePermanentlyAfter: int64(internal_databroker.DefaultDeletePermanentlyAfter.Seconds()),
+	}
+	return db, nil
+}
+
+// Put sets new record for given id with input data.
+func (db *DB) Put(ctx context.Context, id string, data *anypb.Any) error {
+	c := db.pool.Get()
+	defer c.Close()
+	record := db.Get(ctx, id)
+	if record == nil {
+		record = new(databroker.Record)
+		record.CreatedAt = ptypes.TimestampNow()
+	}
+
+	record.Data = data
+	record.ModifiedAt = ptypes.TimestampNow()
+	record.Type = db.recordType
+	record.Id = id
+	record.Version = fmt.Sprintf("%012X", atomic.AddUint64(&db.lastVersion, 1))
+	b, err := json.Marshal(record)
```

Comment:
Using encoding/json with protobuf is not recommended by the protobuf library: https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson?tab=doc

> This package produces a different output than the standard "encoding/json" package, which does not operate correctly on protocol buffer messages.

Reply:
Hmm, when I try switching to `protojson`, the test failed, and somehow hangs.

Reply:
Ah, maybe we should use `proto.Marshal` instead, it's not necessary to use `json` marshaler